realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java

```diff
@@ -370,6 +389,64 @@ public void invokeSchemaChangeListenerIfSchemaChanged() {
         }
     }
 
+    // addCollection(), detachCollections(), reattachCollections() and postToReattachCollections() are used to make
+    // RealmResults stable iterators work. When a Collection is detached from a living OS Results, it won't receive
+    // notifications and its elements won't be changed.
+    // See https://github.com/realm/realm-java/issues/3883 for more information.
+    // Should only be called by Collection's constructor.
+    void addCollection(Collection collection) {
+        if (realmNotifier != null) {
+            collections.add(new WeakReference<Collection>(collection));
+        }
+    }
+
+    // The detaching should happen before transaction begins.
+    private void detachCollections() {
+        for (WeakReference<Collection> collectionRef : collections) {
+            Collection collection = collectionRef.get();
+            if (collection == null) {
+                collections.remove(collectionRef);
+            } else {
+                collection.detach();
+            }
+        }
+    }
+
+    // Ideally the reattaching should happen at the very end of the event loop, but it is impossible for most event
+    // framework. We need to ensure:
+    // 1) It happens before any other coming events get handled (eg: UI redraw event).
+    // 2) It happens before Object Store async callbacks since the Object Store event_loop_signal might use a different
+    //    event queue. This is guaranteed by call this function in the binding_context::before_notify callback.
+    void reattachCollections() {
+        if (isInTransaction()) {
+            // This should never happen.
+            throw new IllegalStateException( "Collection cannot be reattached if the Realm is in transaction." +
+                    " Please remember to commit or cancel transaction before finishing the current event loop.");
+        }
+        for (WeakReference<Collection> collectionRef : collections) {
+            Collection collection = collectionRef.get();
+            if (collection == null) {
+                collections.remove(collectionRef);
+            } else {
+                collection.reattach();
+            }
+        }
+    }
+
+    // To handle the point 1) in the reattachCollections comments.
+    private void postToReattachCollections() {
+        if (realmNotifier != null && !collections.isEmpty() && !reattachCollectionsPosted) {
+            reattachCollectionsPosted = true;
+            realmNotifier.postAtFrontOfQueue(new Runnable() {
+                @Override
+                public void run() {
+                    reattachCollectionsPosted = false;
```

Comment:
maybe use the return boolean of `postAtFrontOfQueue` to update the `reattachCollectionsPosted`